pycbc/cosmology.py

```diff
@@ -23,88 +23,228 @@
 
-class _DistToZ(object):
-    """Class to convert luminosity distance to redshift using the given
-    cosmology.
+DEFAULT_COSMOLOGY = 'Planck15'
+
+
+def get_cosmology(cosmology=None, **kwargs):
+    r"""Gets an astropy cosmology class.
+
+    Parameters
+    ----------
+    cosmology : str or astropy.cosmology.FlatLambdaCDM, optional
+        The name of the cosmology to use. For the list of options, see
+        :py:attr:`astropy.cosmology.parameters.available`. If None, and no
+        other keyword arguments are provided, will default to
+        :py:attr:`DEFAULT_COSMOLOGY`. If an instance of
+        :py:class:`astropy.cosmology.FlatLambdaCDM`, will just return that.
+    \**kwargs :
+        If any other keyword arguments are provided they will be passed to
+        :py:attr:`astropy.cosmology.FlatLambdaCDM` to create a custom
+        cosmology.
+
+    Returns
+    -------
+    astropy.cosmology.FlatLambdaCDM
+        The cosmology to use.
+
+    Examples
+    --------
+    Use the default:
+
+    >>> from pycbc.cosmology import get_cosmology
+    >>> get_cosmology()
+    FlatLambdaCDM(name="Planck15", H0=67.7 km / (Mpc s), Om0=0.307,
+                  Tcmb0=2.725 K, Neff=3.05, m_nu=[0.   0.   0.06] eV,
+                  Ob0=0.0486)
+
+    Don't trust Planck? Use WMAP!
+
+    >>> get_cosmology("WMAP9")
+    FlatLambdaCDM(name="WMAP9", H0=69.3 km / (Mpc s), Om0=0.286, Tcmb0=2.725 K,
+                  Neff=3.04, m_nu=[0. 0. 0.] eV, Ob0=0.0463)
+
+    Don't trust anyone? Create your own!
+
+    >>> get_cosmology(H0=70., Om0=0.3)
+    FlatLambdaCDM(H0=70 km / (Mpc s), Om0=0.3, Tcmb0=0 K, Neff=3.04, m_nu=None,
+                  Ob0=None)
+
+    """
+    if kwargs and cosmology is not None:
+        raise ValueError("if providing custom cosmological parameters, do "
+                         "not provide a `cosmology` argument")
+    if isinstance(cosmology, astropy.cosmology.FlatLambdaCDM):
+        # just return
+        return cosmology
+    if kwargs:
+        cosmology = astropy.cosmology.FlatLambdaCDM(**kwargs)
+    else:
+        if cosmology is None:
+            cosmology = DEFAULT_COSMOLOGY
+        if cosmology not in astropy.cosmology.parameters.available:
+            raise ValueError("unrecognized cosmology {}".format(cosmology))
+        cosmology = getattr(astropy.cosmology, cosmology)
+    return cosmology
+
+
+def z_at_value(func, fval, unit, zmax=1000., **kwargs):
+    r"""Wrapper around astropy.cosmology.z_at_value to handle numpy arrays.
+
+    Getting a z for a cosmological quantity involves numerically inverting
+    ``func``. The ``zmax`` argument sets how large of a z to guess (see
+    :py:func:`astropy.cosmology.z_at_value` for details). If a z is larger than
+    ``zmax``, this will try a larger zmax up to ``zmax * 10**5``. If that still
+    is not large enough, will just return ``numpy.inf``.
+
+    Parameters
+    ----------
+    func : function or method
```

Comment:
i'm curious, what sort of method could this be? presumably something that could be called as e.g. `my_cosmology(z)` and produce a numerical value ..

Reply:
Well, really, any method that takes redshift as input, as the doc says. You'd normally use it with one of the methods of `FlatLambdaCDM` though. This string was copied directly from astropy's documentation for `z_at_value`.